Fix #7168 Remove requests package

cuichenli opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Fixes #7168 . Remove the `requests` package with one module shares the same name. 

## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [ ] The PR has an appropriate "PROJECT: ..." label (Please add this label for the first-pass review of the PR).
- [x] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [ ] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [ ] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

cuichenli committed:
Try requests

cuichenli committed:
Remove dependency of requests

cuichenli committed:
pylint

seanlip commented:
Hi @cuichenli, thanks for your PR. It looks like a bunch of the tests are failing.

Could you please explain a few things:

(a) how you came up with the replacement for the requests file
(b) how you tested this manually to ensure that your code is actually working (it's important to test all functionality manually to ensure that errors don't show up in production)

Note that (b) actually requires investigating the code to understand where the requests library is invoked, and tracing the code paths (or doing some python experiments) to see how to actually invoke it in-context. If you can provide details about that, to help confirm that it is working correctly (and how you know that), that would be great.

Thanks!

cuichenli commented:
Hi, @seanlip for the first question, since the original issue stated that requests should be replaced so I had a look at `requests`'s source code and mocked one similar method, so the main codebase would not need to be changed, not sure is that ok? 

As for the second question, should I add some unit test to address the issue? As I also noticed that `codeclimate/diff-coverage` failed a lot, I assume it is measuring the test coverage? Thanks!

seanlip commented:
For the first question, I think that's fine, but try to keep it as minimal
as possible. I am not sure whether a password etc is needed, for example -
this needs investigation.

For the second question, do a manual test first, since this is one of those
production-type issues where you really do need to be sure it is calling
the other service correctly. Re diff coverage, let me point you to
@lilithxxx for advice since he is working on getting backend code coverage
to 100% -- you should indeed add full unit tests for any new python code.

Thanks!

>
>

cuichenli committed:
Fix existing problems

- sort import
- encode data

cuichenli committed:
Add unit test

cuichenli commented:
Hi @seanlip I have added unit test and manually tried the newly added method (found some problem and fixed). 

In the original codebase, `'api'` is used as username and `feconf.MAILGUN_API_KEY` is used as password (or you can call it token?). So my understanding is it should be fine. Correct me if I'm wrong, thanks! 

https://github.com/oppia/oppia/blob/ba5bfd0b0222de5d2033c3431c7242f3da5edb87/core/platform/email/mailgun_email_services.py#L142

seanlip commented on core/platform/email/requests.py:
```diff
@@ -0,0 +1,48 @@
+    return {'Authorization': auth}
+
+
+def post(server, auth, data):
```
Instead of creating a new module for this, let's just inline a post_to_mailgun() function in mailgun_email_services.py -- no need to pass any args except for `data`. That function can get the mailgun_domain_name, and handle 'api', etc. internally. (That way you can also get rid of a bit of the extra validation above.)

Include, in the function docstring, an explanation that this is adapted from the code for post() in the `requests` library, and provide an Args section (and a Returns section if necessary).

Thanks!

cuichenli replied:
Done in commit `Remove requests module and add inline function post_to_mailgun
`

seanlip reviewed (changes requested):
Thanks @cuichenli! I think your approach looks great, and it's awesome to see the simplification and the removal of the library.

I left a suggestion inline, PTAL!

cuichenli committed:
Remove requests module and add inline function post_to_mailgun

cuichenli commented:
Hi @seanlip I have made update accordingly, however, the pr build failed somehow. Can you help me with this? It looks like I failed some e2e tests while I don't think I have made any changes relates to it. Thanks!
https://travis-ci.org/oppia/oppia/jobs/564915061#L5844

kevinlee12 commented:
Hi @cuichenli, sometimes the e2e tests are flaky so I restarted the tests.

seanlip commented on core/platform/email/mailgun_email_services.py:
```diff
@@ -16,10 +16,32 @@
+
+def post_to_mailgun(data):
+    """Send POST HTTP request to mailgun api. This method is adopted from
+    `requests`'s post method.
```
Sorry, let's make this "the requests library's post method". (Backticks don't really translate well here.)

cuichenli replied:
Done

seanlip commented on core/platform/email/mailgun_email_services.py:
```diff
@@ -16,10 +16,32 @@
+         a file-like object.
+         https://docs.python.org/2/library/urllib2.html
+    """
+    auth_str = 'Basic ' + base64.b64encode(
```
We typically go with interpolation in this codebase, so let's do 'Basic %s' % ...

cuichenli replied:
Done in commit `Move mailgun api check and mailgun domain name check` (Sorry this should have been made to a separate commit.)

seanlip commented on core/platform/email/mailgun_email_services.py:
```diff
@@ -16,10 +16,32 @@
+        - data: dict. The data to be sent in the request's body.
+
+    Returns:
+         a file-like object.
```
Suggest talking a bit about what's (conceptually) contained within the file-like object.

cuichenli replied:
Added in commit `Move mailgun api check and mailgun domain name check`

seanlip reviewed (changes requested):
Hi @cuichenli, looks great, thanks a lot! Just small things left.

seanlip commented:
@kevinlee12 @vojtechjelinek @nithusha21 @DubeySandeep PTAL as codeowners. Thanks!

DubeySandeep commented on core/platform/email/mailgun_email_services.py:
```diff
@@ -16,10 +16,32 @@
 
+import base64
+import urllib
+import urllib2
```
Add a new line here. (Separate local imports from inbuilt libs.)

cuichenli replied:
Done

DubeySandeep commented on core/platform/email/mailgun_email_services.py:
```diff
@@ -112,9 +129,6 @@ def send_bulk_mail(
     if not feconf.MAILGUN_DOMAIN_NAME:
```
Should we move these checks to `post_to_mailgun` and make this method independent of these constants?

cuichenli replied:
Changed in commit `Move mailgun api check and mailgun domain name check`

DubeySandeep commented on core/platform/email/mailgun_email_services.py:
```diff
@@ -55,9 +77,6 @@ def send_mail(
     if not feconf.MAILGUN_DOMAIN_NAME:
```
Ditto*

cuichenli replied:
Done

DubeySandeep reviewed (approved):
Hi @cuichenli, Thanks for the PR, I've left a few minor comments other than that it looks great to me! (Approved!)

DubeySandeep commented:
@cuichenli, Do we also have to remove `requests` from `scripts/backend_tests.py`?
![image](https://user-images.githubusercontent.com/16653571/62077536-f5fec180-b267-11e9-977e-e1428b84e23e.png)

kevinlee12 reviewed (approved):
lgtm (from code owner's)

cuichenli commented:
> @cuichenli, Do we also have to remove `requests` from `scripts/backend_tests.py`?
> ![image](https://user-images.githubusercontent.com/16653571/62077536-f5fec180-b267-11e9-977e-e1428b84e23e.png)

Yes this should be removed, did not notice this naughty guy here. 

I would make changes based on the comments accordingly. 

Also, I noticed that `Codacy/PR Quality Review` check failed, while I can not see what is the problem, can someone help point me to the log so I can fix it? Thanks!

Update: Never mind. Just found the issue.

seanlip commented:
Ah, sorry about that. Please ignore the codacy thing -- we are trying to figure out how to get some code coverage systems working, and this one seems to have some problems. (I've also tried to get rid of it from the repository and I think I've toggled all the switches, so I'm not quite sure why it's still here, sorry!)

seanlip commented:
> Update: Never mind. Just found the issue.

Oh, interesting! Could you describe what the issue was?

cuichenli commented:
@seanlip I just received an email from Codacy which points me to https://app.codacy.com/app/cuichenli/oppia/commit?cid=380358815&bid=13645902&utm_campaign=new_commit&utm_medium=Email&utm_source=Internal&

Basically, it is complaining I used `urlopen` 
![image](https://user-images.githubusercontent.com/29245119/62089577-e0f85180-b2ac-11e9-8671-391f1ae40efb.png)

seanlip commented:
Oh, that's interesting. Thanks @cuichenli! That's good to know.

/cc @brianrodri @lilithxxx FYI

cuichenli committed:
Move mailgun api check and mailgun domain name check

cuichenli committed:
Remove redundant requests module

cuichenli committed:
pylint

cuichenli committed:
Merge branch 'develop' of https://github.com/oppia/oppia into fix-7168

cuichenli commented:
Guys I have made the update based on your review while the E2E test failed again in somewhere I never touched. Can someone please kindly help me to re-trigger the build? Thanks!

DubeySandeep commented:
@cuichenli, I've re-trigger the specific build which was failing unexpectedly. Also, can you please reply to the comments once it's resolved. (As suggested [here](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#address-review-comments-until-all-reviewers-give-lgtm-looks-good-to-me) in the wiki page.)
Thanks! :)

nithusha21 reviewed (approved):
I think this looks great! :).

seanlip commented on core/platform/email/mailgun_email_services.py:
```diff
@@ -16,10 +16,39 @@
+    server = (
+        'https://api.mailgun.net/v3/%s/messages' % feconf.MAILGUN_DOMAIN_NAME)
+    data = urllib.urlencode(data)
+    req = urllib2.Request(server, data, header)   #nosec
```
I think this was due to the codacy issues, right? Just a couple of thoughts:

- If this is what `requests` is already doing, just remove #nosec (we are looking into how to get rid of that check, but don't worry, it won't block submission of this PR).
- On the other hand, if it is actually a security issue, then we should figure out what the issue is and how to fix it.

cuichenli replied:
Yes, this is for the codacy issue. This is one security issue since 
>urlib not only opens http:// or https:// URLs, but also ftp:// and file://. 
https://stackoverflow.com/a/53040523/7676306

The general way to fix this issue is to check the URL format in advance. Since curretnly, we are hardcoding the URL, so I think it should be fine. 

So, I will remove the comment.

Update: Done

seanlip commented on core/platform/email/mailgun_email_services_test.py:
```diff
@@ -16,21 +16,39 @@
 """Tests for the Mailgun API wrapper."""
 
+import urllib2
 from core.platform.email import mailgun_email_services
```
Add newline above this one, to separate stdlib imports from oppia imports.

cuichenli replied:
Done

seanlip commented on core/platform/email/mailgun_email_services_test.py:
```diff
@@ -16,21 +16,39 @@
 class EmailTests(test_utils.GenericTestBase):
     """Tests for sending emails."""
 
+    def test_post_to_mail_gun(self):
```
mailgun is one word.

cuichenli replied:
Done

seanlip reviewed (approved):
Thanks so much, @cuichenli! Just a couple tiny things left, then we can merge this. (Please see inline comments.)

Also, congrats on your first PR to Oppia! :D

cuichenli committed:
Remove #nosec and format

seanlip reviewed (approved):
All good. Thanks @cuichenli! Full LGTM from my end. 

@vojtechjelinek, could you PTAL so that this can be merged? Thanks!

vojtechjelinek reviewed (approved):
LGTM!
